src/Internal/Plutip/Spawn.purs

```diff
@@ -86,15 +86,22 @@ killOnExit child = do
       kill SIGINT child
 
 foreign import _rmdirSync :: FilePath -> Effect Unit
+foreign import _mkdirSync :: FilePath -> Effect Unit
 
 cleanupTmpDir :: ChildProcess -> FilePath -> FilePath -> Effect Unit
-cleanupTmpDir child workingDir testClusterDir = do
+cleanupTmpDir child workingDir rootDir = do
   ChildProcess.onExit child \_ -> do
     _rmdirSync workingDir
   -- TODO
   -- cleanup directories in case of ctrl+c exit,
   -- code below doesn't work as intended.
   -- https://github.com/Plutonomicon/cardano-transaction-lib/issues/1176
-  Process.onExit \_ -> do
-    _rmdirSync testClusterDir
-    _rmdirSync workingDir
+  -- Process.onSignal SIGINT do
```

Comment:
This would be the cleanest solution. The problem is though that after registering the `onSignal SIGINT` listener, the `testClusterDir` and the `workingDir` change with each child process. So this handler tries to remove directories that have already been removed by the `ChildProcess.onExit` listener on line 93.

To solve this, we would need to know which are the current `testClusterDir` and `workingDir` when ctrl+c is pressed. But as far as I can see that's impossible to know at the moment of registering the listener.

Reply:
If they have already been deleted, just do nothing.

Reply:
the problem is that it always tries to delete the tmp directories from the first child process (as that is the moment when the listener is created). but it doesn't delete the tmp directories created by the current (i.e when ctrl+c is pressed) child process

Reply:
Isn't `cleanupTmpDir` called for each of the child processes? If so, why the correct tmp dirs are not passed?

Reply:
I'm stuck here. The problem from above is solved, the problem was that by using `Process.exit` in the SIGINT listener, the listeners registered in the second and following child processes never gets executed. 

so we must call `Process.exit` only in the last listener, which I'm trying to do in `cleanupOnSigint`. TTriggering SIGINT produces the following error though which I just can't figure out. Do I maybe have to work in Aff instead of Effect? 

  Error: (ClientHttpError There was a problem making the request: request failed)
    at Object.exports.error (/home/pekka/ctl/cardano-transaction-lib/output/Effect.Exception/foreign.js:8:10)
    at Object.$$throw [as throw] (/home/pekka/ctl/cardano-transaction-lib/output/Effect.Exception/index.js:18:45)
    at /home/pekka/ctl/cardano-transaction-lib/output/Ctl.Internal.Plutip.Server/index.js:151:57
    at /home/pekka/ctl/cardano-transaction-lib/output/Data.Either/index.js:203:24
    at /home/pekka/ctl/cardano-transaction-lib/output/Ctl.Internal.Plutip.Server/index.js:153:71
    at run (/home/pekka/ctl/cardano-transaction-lib/output/Effect.Aff/foreign.js:278:22)
    at /home/pekka/ctl/cardano-transaction-lib/output/Effect.Aff/foreign.js:348:19
    at drain (/home/pekka/ctl/cardano-transaction-lib/output/Effect.Aff/foreign.js:120:9)
    at Object.enqueue (/home/pekka/ctl/cardano-transaction-lib/output/Effect.Aff/foreign.js:141:11)
    at /home/pekka/ctl/cardano-transaction-lib/output/Effect.Aff/foreign.js:339:27

Reply:
Some thoughts: multiple listeners conflicting can be fixed by detatching the listener when the resources are cleaned up in the bracket. This would also be a good time to run the cleanup.
It sounds like maybe

I'm surprised you get that error after calling exit, exit should shut down the process as quickly as possible without allowing other events to be run (other than those that listen for exit).

Reply:
Also I'm curious if you have noticed the behaviour with ctrl c and spago like I have in my comment below

Reply:
thanks for the input. I think detaching the `onSignal SIGINT` listener after each child process is exactly what I need but I couldn't figure out how to do it. would I need the ffi, or something similar to this: https://stackoverflow.com/questions/57259201/how-to-schedule-a-removeeventlistener-inside-an-addeventlistener ?

Reply:
about the error: I think it occurs before `Process.exit`, but I couldn't yet figure out where exactly. 

ctrl-c for me doesn't behave as you describe below. The `onSignal SIGINT` handler gets triggered and that wasn't the case for you if I get it right.

Reply:
I guess really we shouldn't detach listeners (by name) in purescript because of referential transparency, we'd need to create a version of onSignal which returns the function it attached, so we can later detach.

```purescript
foreign import data OnSignalRef :: Type

foreign import onSignalImpl :: String -> Effect Unit -> Effect OnSignalRef

foreign import removeOnSignal :: OnSignalRef -> Effect Unit

onSignal :: Signal -> Effect Unit -> Effect OnSignalRef
onSignal sig = onSignalImpl (Signal.toString sig)
```

```javascript
exports.onSignalImpl = signal => callback => () => {
  process.on(signal, callback);
  return { signal, callback };
};

exports.removeOnSignal = { signal, callback } => () => {
  process.removeListener(signal, callback);
};
```

This also helps not make https://github.com/Plutonomicon/cardano-transaction-lib/issues/864 worse by making sure we cleanup

Reply:
Fully agree @jy14898

Reply:
We can also make the FFI function that attaches a handler return a callback that de-attaches it.